src/ChannelServer/World/Entities/Components/Buffs.cs

```diff
@@ -55,12 +55,15 @@ public void AddSilent(Buff buff)
 		/// <param name="buff"></param>
 		public void Add(Buff buff)
 		{
-			var buffExists = _buffs.ContainsKey(buff.Id);
-			this.AddSilent(buff);
-			if (buffExists)
-				Send.ZC_BUFF_UPDATE(this.Character, buff);
-			else
-				Send.ZC_BUFF_ADD(this.Character, buff);
+			var buffExists = this.Has(buff.Id);
+			lock (_buffs)
```

Comment:
This lock is unnecessary, because you switched to using Has and AddSilent, which both handle the locking. You should also only lock for as long as necessary, meaning as long as you need to access the resource that is to be locked, because while you're sending packets, no other thread can access the locked area.

Reply:
What's the preferred way to do it? Use the lock then access the resource and send the packet within the lock or lock it to get the reference and send packet outside of the lock?

Reply:
Like I said, only lock for as long as absolutely necessary, to free up the resource for other threads. In this case, and if you weren't using Has and AddSilent, which handle it, you might do something like this:
```cs
public void Add(Buff buff)
{
	var updated = false;
	lock (_buffs)
	{
		updated = _buffs.ContainsKey(buff.Id);
		_buffs[buff.Id] = buff;
	}

	if(updated)
		Send.ZC_BUFF_UPDATE(this.Owner, buff);
	else
		Send.ZC_BUFF_ADD(this.Owner, buff);
}
```